From e687fb52b11450c6128657022281a2961d32aac9 Mon Sep 17 00:00:00 2001
From: Lawrence Rust <lvr@softsystem.co.uk>
Date: Fri, 10 Jan 2014 17:17:07 +0000
Subject: [PATCH 198/333] Player: Avoid unnecessary calls to ReinitOSD

When changing channel to one on the same mux the signal tuning OSD
often is not shown because a change in aspect ratio is detected
after the OSD message 'on known multiplex' is received.

This patch adds the aspect ratio to MythPlayer::SetVideoParams
so that the OSD is not reinitialised unncessarily.

Signed-off-by: Lawrence Rust <lvr@softsystem.co.uk>
---
 mythtv/libs/libmythtv/avformatdecoder.cpp |    6 +++---
 mythtv/libs/libmythtv/mythplayer.cpp      |    9 +++++++--
 mythtv/libs/libmythtv/mythplayer.h        |    2 +-
 3 files changed, 11 insertions(+), 6 deletions(-)

diff --git a/mythtv/libs/libmythtv/avformatdecoder.cpp b/mythtv/libs/libmythtv/avformatdecoder.cpp
index 5a66c54..502bd3c 100644
--- a/mythtv/libs/libmythtv/avformatdecoder.cpp
+++ b/mythtv/libs/libmythtv/avformatdecoder.cpp
@@ -1680,7 +1680,7 @@ void AvFormatDecoder::InitVideoCodec(AVStream *stream, AVCodecContext *enc,
         }
 
         m_parent->SetKeyframeDistance(keyframedist);
-        m_parent->SetVideoParams(width, height, fps, kScan_Detect);
+        m_parent->SetVideoParams(width, height, fps, current_aspect, kScan_Detect);
         if (LCD *lcd = LCD::Get())
         {
             LCDVideoFormatSet video_format;
@@ -3308,7 +3308,7 @@ void AvFormatDecoder::MpegPreProcessPkt(AVStream *stream, AVPacket *pkt)
 
             if (changed)
             {
-                m_parent->SetVideoParams(width, height, seqFPS, kScan_Detect);
+                m_parent->SetVideoParams(width, height, seqFPS, current_aspect, kScan_Detect);
 
                 current_width  = width;
                 current_height = height;
@@ -3415,7 +3415,7 @@ int AvFormatDecoder::H264PreProcessPkt(AVStream *stream, AVPacket *pkt)
 
         if (fps_changed || res_changed)
         {
-            m_parent->SetVideoParams(width, height, seqFPS, kScan_Detect);
+            m_parent->SetVideoParams(width, height, seqFPS, current_aspect, kScan_Detect);
 
             current_width  = width;
             current_height = height;
diff --git a/mythtv/libs/libmythtv/mythplayer.cpp b/mythtv/libs/libmythtv/mythplayer.cpp
index 45d996a..5130425 100644
--- a/mythtv/libs/libmythtv/mythplayer.cpp
+++ b/mythtv/libs/libmythtv/mythplayer.cpp
@@ -808,7 +808,7 @@ void MythPlayer::SetScanType(FrameScanType scan)
     m_scan = scan;
 }
 
-void MythPlayer::SetVideoParams(int width, int height, double fps,
+void MythPlayer::SetVideoParams(int width, int height, double fps, float aspect,
                                 FrameScanType scan)
 {
     bool paramsChanged = false;
@@ -818,7 +818,12 @@ void MythPlayer::SetVideoParams(int width, int height, double fps,
         paramsChanged  = true;
         video_dim      = QSize((width + 15) & ~0xf, (height + 15) & ~0xf);
         video_disp_dim = QSize(width, height);
-        video_aspect   = (float)width / height;
+    }
+
+    if (aspect > 0.f)
+    {
+        paramsChanged  = true;
+        video_aspect = aspect;
     }
 
     if (!qIsNaN(fps) && fps > 0.0f && fps < 121.0f)
diff --git a/mythtv/libs/libmythtv/mythplayer.h b/mythtv/libs/libmythtv/mythplayer.h
index 0a0be92..87ec90f 100644
--- a/mythtv/libs/libmythtv/mythplayer.h
+++ b/mythtv/libs/libmythtv/mythplayer.h
@@ -159,7 +159,7 @@ class MTV_PUBLIC MythPlayer
     void SetWatchingRecording(bool mode);
     void SetWatched(bool forceWatched = false);
     void SetKeyframeDistance(int keyframedistance);
-    void SetVideoParams(int w, int h, double fps,
+    void SetVideoParams(int w, int h, double fps, float aspect = 0,
                         FrameScanType scan = kScan_Ignore);
     void SetFileLength(int total, int frames);
     void SetDuration(int duration);
-- 
1.7.9.5

